void_migration/void_migration.py

```diff
@@ -466,6 +516,156 @@ def close_voids(u, v, s):
+
+            # mass_i.append(np.sum(np.array(mass_k)[~np.isnan(np.array(mass_k))]))   ##  sum of non-nan masses
+            mass_i.append(np.sum(mass_k))   
+        mass_fun.append(np.sum(mass_i))
+
+    return mass_fun
+
+
+def charge_discharge(p,op_arr,t,change_s_ms):
+
+    '''
+    As of now two times (t_fill and t_empty) are calculated and t_settle is given
+    t_fill - filling time
+    t_settle - allow the cells to settle
+    t_empty - end time
+    for mono-disperse conditions, increase very slightly the grainsize for different cycles 
+    (mainly to assign different colors for different cycle), and 
+    this is done in the time_march function 
+    '''
+    
+    s_ms = 0
+    if p.gsd_mode == 'mono':
+
+        colsm = []
+        for i in range(p.no_of_cycles):
+            if i % 2 == 0:
+                colsm.append(change_s_ms[0])
+            else:
+                colsm.append(change_s_ms[1])
+
+    res = [sub['t_empty'] for sub in op_arr]  #Just take the t_empty values from the op_arr dictionary
+    
+    yj = []
+    for j in range(len(res)):
+        if t < int(np.ceil(res[j]/p.dt)):    
+            yj.append(op_arr[j])
+    
+    if p.gsd_mode == 'mono':
+        col_index = colsm[int(len(op_arr) - len(yj))]
+
+
+    if t <= int(np.ceil(yj[0].get('t_fill')/p.dt)):   ##To always pick the time belonging to the first element in the array 
+        if p.gsd_mode == 'mono':
+            p.s_m = col_index
+            p.add_voids = 'place_on_top' #'pour_base'
+        else:
+            p.add_voids = 'place_on_top'
+
+    elif int(np.ceil(yj[0].get('t_fill')/p.dt)) < t <= int(np.ceil(yj[0].get('t_settle')/p.dt)):
+        p.add_voids = 'None'
+
+    elif int(np.ceil(yj[0].get('t_settle')/p.dt)) < t < int(np.ceil(yj[0].get('t_empty')/p.dt)-1):
+        p.add_voids = "central_outlet"
+        p.save_outlet = True
+
+    return p,change_s_ms
+
+
+def cals_in_cd(p):
+    '''
+    Calculates the time required for filling, settling and emptying for each cycle and a desired number of cycles
+    M_total : Total mass for a given solid fraction
+    Mass_in_per_u_t : Mass going in per unit time
+    Mass_out_per_u_t : Mass coming out of silo per unit time 
+    T_in : filling time calculated based on M_total, Mass_in_per_u_t and p.dt
+    T_settle : settling time 
+    T_out : emptying time calculated based on M_in, Mass_out_per_u_t and p.dt
+    op_arr : holds the values for t_fill, t_settle and t_empty for each cycle and for a desired number of cycles  
+
+    '''
+    if p.gsd_mode == 'mono':
+        v_sphere = (4/3) * np.pi * ((p.s_m + p.s_M)/2/2)**3   ## m^3
+    elif p.gsd_mode == 'bi':
+        v_sphere = (((4/3)*np.pi*(p.s_m/2)**3) + ((4/3)*np.pi*(p.s_M/2)**3)) / 2
+    else:
+        v_sphere = (4/3) * np.pi * ((p.s_m + p.s_M)/2/2)**3   ## m^3
+
+    rho_p = 3950    ## kg/m^3
+
+    # Total mass
+    M_total = p.nx * p.ny * p.nm * v_sphere * rho_p * p.nu_cs
```

Comment:
is this the total mass in the system? if so it is incorrect. the total mass is just the number of cells filled with material (i.e. non-nan) times dx times dy times the density (in kg/m^2?). If you want to define the density in kg/m^3 then we need a nominal thickness of the cells.